src/proto/beats.proto

```diff
@@ -10,22 +11,30 @@ enum Source {
   USER = 2;
 }
 
-message Beat {
-  optional int32 frame_position = 1;
-  optional bool enabled = 2 [ default = true ];
-  optional Source source = 3 [ default = ANALYZER ];
+enum Type {
+  BEAT = 0;
+  BAR = 1;
+  PHRASE = 2;
+}
+
+message TimeSignature {
+  required int32 beats_per_bar = 1 [ default = 4];
```

Comment:
What is the solution for single exceptional bars with 4 1/2 or 3 1/2 beats.

Reply:
I am not sure I got that question.
If a bar has 4 1/2 beats it would mean the downbeat for the bar specifies a time signature as TimeSignature(4,2)

Reply:
@daschuer Please let's only consider simple time signatures with an integer *note value* aka *beat unit*. This is sufficient for our use case. [Fractional beats](https://en.wikipedia.org/wiki/Time_signature#Variants) are out of scope.

Reply:
This was a real question! 

I don't think we necessarily required to introduce floats here, we just to need to be aware what to do, 
We can have something like this: 
```
D       b       b       b   D       b       b       b      D       b       b       b       D       b       b       b  
```
That is not uncommon. It can be intended or a Drummer hick up.  

How would this be described in the protobuf. 

The other thing I struggle with are Metal tracks where 7/16 or such is not uncommon. 
I am in doubt that we would like to have out beat grid cluttered with every 1/16 beats, because only two of them are notable as beats and probably found by the beat detector. It is just that one of them is slightly off
(@crisclacerda please verify.) 

A 4/4 measure can be counted as 16/16 measure as  

```
1 e + a 2 e + a 3 e + a 4 e + a 1 e + a 2 e + a 3 e + a 4 e + a 
D       b       b       b       D       b       b       b  
```

So the 7/16 example will be 
```
1 e + a 2 e + 1 e + a 2 e + 1 e + a 2 e + 1 e + a 2 e + 
D       b     D       b     D       b     D       b  
```

This video shows a 15/16 measure which is even more weird 
https://www.youtube.com/watch?v=GwCyIZ-v9CA

How would we express this?

Reply:
> That is not uncommon. It can be intended or a Drummer hick up.

In the new encoding scheme (to be developed) using beat markers, we won't store frame positions in the protobuf at all.
Instead, a beat marker will be used which is represented as:
```
BeatMarker {
    required int beatIndex
    optional Bpm bpm
    optional TimeSignature timeSignature
    optional Section section
    optional Phrase phrase
}

Beats {
    required firstBeatFramePosition
    repeated BeatMarker beatMarkers
}
```

A beat position is determined by previous beat position and its BPM on the basis of their indices.

So if a track is jittery, there is a beat marker at every index, but if the track has uniform bpm, there is only one BeatMarker at the beginning of the track.

This scheme is a hybrid between BeatMap and BeatGrid.

Reply:
OK, But how are the two examples expressed than? What is the signature Mixxx Displays during the 3.5/4 measure? 
How to visualize x/16 measures. The waveform will be white for beatmarkers and 1/16 markers have no use-case.

Reply:
A 3.5/4 measure could also be expressed as 7/8. Alternatively, we can keep 4 beats in the bar but let the user cut the last beat short without moving the positions of the other beats.

Reply:
Yes, it is a 7/8 bar.
Without additional thinking this will be visualized as before of BPM as before.
I don't think we want that. 

I have the feeling we need to go back to the cutting table. What is the use case of the signature info? 

I am afraid we combine here two related things that might be kept separate or at least considered separate. The sparse representation and user information.

1. The sparse representation of const beatgrid. We can give the position of the first downbeat, the BPM and the number of beats per bar.
In this case we just need to re-adjust the early down-beat and we are done.

2. Inform the User about irregular measures. The user should be warned not to use beatmatching across this irregular meassure. How do we want to give that to a user? 

3. Display the main signature in the library.

4. The hardest nut is the 7/16 Mattle track.
What do you suggest with this? I don't want to have 1/16 Beats on the waveform. I don't want to set every beat manually.
I may want to sync the track using the down-beat only. I may want to doo loop tricks. 
How about defining the first as creazy a as it is and allow to repeat it until the next measure with a different signature.

What do you think?

Reply:
> Without additional thinking this will be visualized as double BPM as before.
I don't think we want that.

I agree. I am thinking the best solution would be not assuming that every beat in a bar is an equal length of time. If we do not assume that, a 3.5 beat measure could be a 4/4 measure with the last beat shorter than the others.

Reply:
We never assumed that. In fact, time signature and beat track (the more general term for beat map) are quite independent.
A time signature only encodes the information about the downbeats.
Beat length is determined by a preceding beat marker that marks a change of BPM.

Let's look at a series of Beat Markers.

```js
[
{
  index: 0
  ts: 4/4
  bpm: 60
},
{
  index: 3
  bpm: 120
},
{
  index: 4
  bpm: 60 // restored for the next bar
},
// The pattern repeats
]
```

Reply:
Yes, a 4/4 measure in this case does make sense from the sparse representation view however it is just wrong as information for the user. 

Do we have a plan how to the the info to the user? I can think of something like the key label that shows the time signature. In our case it will change from 4/4 to 5/8 or "?"  if you like. This solves not the odd measure warning use case though.  

How about integrate Bars into the protobuf

```
message Bar {
  // The frame position should always be explicitly set
  // otherwise serialization would lead to SIGABRT.
  required double frame_position = 1 [ default = 0 ];
  optional Source source = 3 [ default = ANALYZER ];
  optional string phrase =  6 [ default = "intro"];  // not empty = phrase start 
  optional string signature = 5 [ default = "4/4"]; // not empty = down beat 
  optional int number_of_Beats = 6 [default = 4]; 
  repeated beat_frame_position_offset = 7; // can be empty for a const bar.  
}
```

This would allow to define one complex bar, call freely give it a signature name and repeat it until a new one is defined. 
It can also be stripped down to be only a beat. This is the case when only frame_position is set.

Reply:
Unfortunately the protobuf proposed in this PR does not work for me. There are too many degrees of freedom and it is unhandy for a smooth manual editing workflow.

This is the wiki page that summarizes the state of discussion:
https://github.com/mixxxdj/mixxx/wiki/Beat-and-Bar-Edit-Workflow

I have suggested to use this buffer:

```
[
  frame_position = 0;
  source = USER;
  signature_nominator = 7; 
  signature_denominator = 16;  
  quarter_beats[1] = 4; // for 4/16
]  
[
  frame_position = 80000;
  source = USER;
  signature_nominator = -1; // REPEAT
]  
[
  frame_position = 375000;
  source = USER;
  signature_nominator = -2; // END
]    
```
Will it work for you? 
What do we need to change.

Reply:
In the long term, I am planning to use Beat Indices to reference beats rather than frame positions. No change needed for this PR.

Reply:
I have another suggestion. Can we defer beatgrid editing features to another PR and strictly focus on the removal of BeatMap and BeatGrid in this PR?
We won't have any beatgrid editing UI in this PR anyways since it is being developed in #2844.
So I am only majorly focusing on correctly refactoring the current code and make everything work like it did with BeatMap and BeatGrid.

Reply:
This PR will use the new protobuffer, right? I think it is a good idea to switch to one that is a base for new features and not switch it again in a follow up PR.  We must not merge a half baked solution especially if other PRs will rely on the new format. 

Please tell us more about your beat indices idea. It sounds interesting.

Reply:
A part of it has been implemented in #2844. We are using Beat Index not Frame Positions to refer to beats while rendering and setting a particular beat as a downbeat via the context menu.
Storing the same can be possible when we are storing the BPM and not frame position.
This is like a beat grid which stores the first position only and the rest can be calculated with BPM.
In a track with multiple/variable BPM, just store the first beat position and a series of beat markers with [ beat index + bpm ] to find other beats. The extreme case would be when all beats are jittery and we store a beat marker for every beat.

Reply:
So the user will give the tempo as BPM? 

I think we have two use cases, for electronic tracks with integer BPMs this might work. 
Unfortunately not for live tracks. 

I am also in doubt that he can do this precise enough. He will likely do decimal rounding which leads to a increasing offset. 

I think the us will point a beat. 

Do you think the idea with the snap grid https://github.com/mixxxdj/mixxx/pull/2917 will work? In this case it is precise by 11 ms, the detector step width. 
@crisclacerda smoothing code can than also used to the user edited beats to partition the grid into constant regions. 

I think it is reasonable have frames in the proto buffer as well. Because the engine code works with frames and that releases us from calculating the frames every time the beat-grid is loaded into the memory.